### DISCUSSION TITLE
How to get the current song playing?

### DISCUSSION THREAD

**Discussion by sanderfoobar** [savonet/liquidsoap#3869] [General]
How to ask Liquidsoap for the current **filename** (or path) of the song that is playing?

Previously, I used the command `<radio_name>.metadata` via telnet for this:

```text
set("server.telnet", true)
set("server.telnet.bind_addr", "127.0.0.1")
set("server.telnet.port", 7555)
set("server.telnet.reverse_dns", false)
```

```text
$ nc 127.0.0.1 7555
radio.metadata
```

And it would output a list of songs that were played and is playing, which includes a `filename=` key that I parsed. However, after upgrading to Liquidsoap 2.2.4, the telnet output has been changed, and I no longer can find `filename=` in the output.

For the time being, I am parsing `journalctl -xefu liquidsoap` and watching [for a specific line](https://github.com/savonet/liquidsoap/blob/a30127e5f38af46a36c1aeca217086723c3982bf/src/core/sources/request_dynamic.ml#L109), since that mentions an absolute filepath. While that works, it is very hacky.

Note that in my case, querying icecast2 is not sufficient, as that only shows file metadata like artist/title. I need the exact filename or filepath.

Any advice welcome.

**Comment by gAlleb** [comment]
Hi @sanderfoobar! You can always write your own telnet command, really.

After your source (say `radio`) has been defined paste the following:

```ruby
def get_filename(_)
  
  m = radio.last_metadata() ?? []

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end

  filename = null_list("filename", m)

  "filename=#{filename}"

end
server.register("show_me_filename_por_favor", get_filename)
```

Now when you enter `show_me_filename_por_favor` into your telnet terminal - you will get `filename=/path/to/file.flac` of current file being played.

Moreover, you can register simple http sever for this command. Add `harbor.http.register.simple(port=8008, method="GET", "/show_me_filename_por_favor", get_filename)` after `server.register("...` and that's it. 

Now if you go to `http://localhost:8008/show_me_filename_por_favor` you'd get the same `filename=/path/to/file.flac` response. 

If you need filename with `""` just change this line `"filename=#{filename}"` into `'filename="#{filename}"'`.

**Comment by sanderfoobar** [comment]
That works great, thanks.
